services/medco/api.go

```diff
@@ -0,0 +1,95 @@
+package medco
+
+import (
+	"github.com/btcsuite/goleveldb/leveldb/errors"
+	"github.com/dedis/cothority/log"
+	"github.com/dedis/cothority/network"
+	"github.com/dedis/cothority/sda"
+	"github.com/dedis/cothority/services/medco/libmedco"
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/config"
+	"strconv"
+)
+
+// API represents a client with the server to which he is connected and its public/private key pair.
+type API struct {
+	*sda.Client
+	entryPoint        *network.ServerIdentity
+	localClientNumber int64
+	public            abstract.Point
+	private           abstract.Scalar
+}
+
+var localClientCounter = int64(0)
+
+// NewMedcoClient constructor of a client.
+func NewMedcoClient(entryPoint *network.ServerIdentity) *API {
+	keys := config.NewKeyPair(network.Suite)
+	newClient := &API{
+		Client:            sda.NewClient(ServiceName),
+		entryPoint:        entryPoint,
+		localClientNumber: localClientCounter,
+		public:            keys.Public,
+		private:           keys.Secret,
+	}
+
+	localClientCounter++
+	return newClient
+}
+
+// CreateSurvey creates a survey based on a set of entities (servers) and a survey description.
+func (c *API) CreateSurvey(entities *sda.Roster, surveyDescription libmedco.SurveyDescription) (*libmedco.SurveyID, error) {
+	log.Lvl1(c, "is creating a survey.")
+	resp, err := c.Send(c.entryPoint, &SurveyCreationQuery{nil, *entities, surveyDescription})
+	if err != nil {
+		log.Error("Got error when creating survey: " + err.Error())
+		return nil, err
+	}
+	log.Lvl1(c, "successfully created the survey with ID", resp.Msg.(ServiceResponse).SurveyID)
+	surveyID := resp.Msg.(ServiceResponse).SurveyID
+	return &surveyID, nil
+}
+
+// SendSurveyResultsData creates and sends a client response encrypted with the cothority collective key.
+func (c *API) SendSurveyResultsData(surveyID libmedco.SurveyID, grouping, aggregating []int64, groupKey abstract.Point) error {
+	log.Lvl1(c, "responds {", grouping, ",", aggregating, "}")
+	encGrouping := libmedco.EncryptIntVector(groupKey, grouping)
+	encAggregating := libmedco.EncryptIntVector(groupKey, aggregating)
+	_, err := c.Send(c.entryPoint, &SurveyResponseQuery{surveyID,
+		libmedco.ClientResponse{
+			ProbabilisticGroupingAttributes: *encGrouping,
+			AggregatingAttributes:           *encAggregating}})
+	if err != nil {
+		log.Error("Got error when sending a message: " + err.Error())
```

Comment:
We only use `log.Error` if the error is not sent to the caller. Else we let the caller handle the error and decide if he wants to call `log.Error` (if he can't send up the error) or ignore it.